File: py_code/简单题/LC_2105.py
# https://leetcode.cn/problems/watering-plants-ii/description/

'''
双指针模拟
'''
class Solution:
    def minimumRefill(self, plants: List[int], capacityA: int, capacityB: int) -> int:
        n = len(plants)
        i, j = 0, n - 1
        v1, v2 = capacityA, capacityB
        ans = 0
        while i <= j:
            if i == j:
                if v1 == v2 or v1 > v2:
                    ans += v1 < plants[i]
                else:
                    ans += v2 < plants[i]
                break

            if v1 < plants[i]:
                ans += 1
                v1 = capacityA - plants[i]
            else:
                v1 -= plants[i]

            if v2 < plants[j]:
                ans += 1
                v2 = capacityB - plants[j]
            else:
                v2 -= plants[j]

            i += 1
            j -= 1
        return ans

File: leetcode/src/com/fqh/双指针/LC_2105.java
package com.fqh.双指针;

/**
 * @author ikun
 * @version v1.0.0
 * @since 2024/5/9 22:52
 **/
public class LC_2105 {

    // https://leetcode.cn/problems/watering-plants-ii/
    public int minimumRefill(int[] plants, int capacityA, int capacityB) {
        int n = plants.length;
        int a = capacityA, b = capacityB;
        int i = 0, j = n - 1;
        int ans = 0;
        while (i <= j) {
            if (i == j) {
                if (a >= b) ans += (a < plants[i] ? 1 : 0);
                else ans += (b < plants[j] ? 1 : 0);
                break;
            }
            if (a >= plants[i]) {
                a -= plants[i];
            } else {
                ans++;
                a = capacityA - plants[i];
            }

            if (b >= plants[j]) {
                b -= plants[j];
            } else {
                ans++;
                b = capacityB - plants[j];
            }
            i++; j--;
        }
        return ans;
    }
}
